Blinding/src/DisplaceDigiTimes_module.cc

```diff
@@ -0,0 +1,211 @@
+      std::string msg = "ambiguous KalSeedCollection (length " + std::to_string(kalseeds_handle->size()) + "), cannot assign reference time";
+      throw cet::exception("DisplaceDigiTimes") << msg << std::endl;
+    }
+    auto kalseed = (*kalseeds_handle)[0];
+
+    // search for downward-going KalIntersection with a matching surface
+    KalIntersection intersection;
+    auto sit = _surface_ids.begin();
+    bool adequate = false;
+    while ((!adequate) && (sit != _surface_ids.end())){
+      const auto& iits = kalseed.intersections(*sit);
+      // there are at most 2 valid intersections (upgoing / downgoing)
+      // so selecting on which one is downgoing is sufficient
+      for (const auto& iit: iits){
+        auto direction = iit->momentum3().Unit().Z();
+        if (0 < direction){
+          intersection = *iit;
+          adequate = true;
+        }
+      }
+      sit++;
+    }
+
+    // if no valid intersection found, then we cannot use this event
+    if (!adequate){
+      std::string msg = "no valid KalIntersection found";
+      throw cet::exception("DisplaceDigiTimes") << msg << std::endl;
+    }
+    // otherwise, take the time of the intersection as the reference time
+    double reference_time = intersection.time();
+    double target_time = this->sample_target_time();
+    double shift = target_time - reference_time;
+
+    // apply the shift to tracker digis
+    const auto& electronics = _tracker_conditions_handle.get(event.id());
+    auto digis = std::make_unique<StrawDigiCollection>();
+    auto adcss = std::make_unique<StrawDigiADCWaveformCollection>();
+    if (0 < digi_handle->size()){
+      for (const auto& old: *digi_handle){
+        auto sid = old.strawId();
+        auto tdc = old.TDC();
+        auto tot = old.TOT();
+        auto pmp = old.PMP();
+
+        // displace tdc times
+        for (size_t i = 0 ; i < TrkTypes::NENDS ; i++){
+          auto analog = this->straw_digital_to_analog(tdc[i], electronics);
+          auto shifted = analog + shift;
+          tdc[i] = this->straw_analog_to_digital(shifted, electronics);
+        }
+
+        // new digi is identical to input digi, with displaced tdc
+        digis->emplace_back(sid, tdc, tot, pmp);
+      }
+
+      for (const auto& old: *adcs_handle){
+        adcss->push_back(old);
+      }
+    }
+    event.put(std::move(digis));
+    event.put(std::move(adcss));
+
+    // calorimeter...
+  }
+
+  double DisplaceDigiTimes::sample_target_time(){
+    const double scaled = _target_sampler->fire();
+    const double rv = _target_distribution.sample(scaled);
+    return rv;
+  }
+
+  TrkTypes::TDCValue DisplaceDigiTimes::straw_analog_to_digital(double time, const StrawElectronics& electronics){
+    auto rv = electronics.tdcResponse(time);
+    return rv;
+  }
+
+  double DisplaceDigiTimes::straw_digital_to_analog(TrkTypes::TDCValue time, const StrawElectronics& electronics){
+    auto rv = static_cast<double>(time) * electronics.tdcLSB();
```

Comment:
should this function be moved to StrawElectronics?
Are there edge effects one needs to worry about? How about the channel-specific effects described [here](url)?

Reply:
I don't feel strongly about shifting this into the electronics class or not. If there is any other use case for inverting (IMO, likely) then it would make sense to. Maybe @bonventre can comment on nuance that may be missing --- I just didn't see anything else in the existing electronics utility to make use of.

Reply:
Right now this is missing _electronicsTimeDelay from strawelectronics. I think this is a redundant parameter that is set to zero and can be removed, but for now maybe its easier to put this function in strawelectronics and add this offset back in there.